### PR TITLE
Fixed ground to image and image to ground not giving inverse results.

### PR DESCRIPTION
**Pull request by jessemapel**
This also adds a very simple test to the line scan camera and some test stubs.

### CODE REVIEW

File: src/UsgsAstroLsSensorModel.cpp [hunk 1]
```diff
@@ -561,57 +528,89 @@ csm::ImageCoord UsgsAstroLsSensorModel::groundToImage(
...
+   double lastOffset = computeViewingPixel(lastTime, ground_pt, adj, pixelPrec/2).line - 0.5;
+
+   // Check if both offsets have the same sign.
+   // This means there is not guaranteed to be a zero.
+   if ((firstOffset > 0) != (lastOffset < 0)) {
+        throw csm::Warning(
+           csm::Warning::IMAGE_COORD_OUT_OF_BOUNDS,
+           "The image coordinate is out of bounds of the image size.",
+           "UsgsAstroLsSensorModel::groundToImage");
+   }
```
Comment (jessemapel) [inline src/UsgsAstroLsSensorModel.cpp]: This may need to go away? I think that switching to the secant method means we don't need opposite signs.

File: src/UsgsAstroLsSensorModel.cpp [hunk 2]
```diff
@@ -2766,7 +2773,7 @@ std::string UsgsAstroLsSensorModel::constructStateFromIsd(const std::string imag
 
 
    state["m_parameterVals"] = std::vector<double>(NUM_PARAMETERS, 0.0);
-   state["m_parameterVals"][15] = state["m_focal"].get<double>();
+   // state["m_parameterVals"][15] = state["m_focal"].get<double>();
```
Comment (jessemapel) [inline src/UsgsAstroLsSensorModel.cpp]: We are storing focal length in two places. We should either only store it in one place or only use on in calculations, see line 1733.

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -3,7 +3,8 @@ cmake_minimum_required(VERSION 3.10)
 # Link runCSMCameraModelTests with what we want to test and the GTest and pthread library
 add_executable(runCSMCameraModelTests
                PluginTests.cpp
-               FrameCameraTests.cpp)
+               FrameCameraTests.cpp
+               LineScanCameraTests.cpp)
```
Comment (jessemapel) [inline tests/CMakeLists.txt]: We should switch to what ISIS does and just glob all of the cpp files in the tests directory.

File: tests/data/constVelocityLineScan.json [hunk 4]
```diff
@@ -3,15 +3,15 @@
...
-    [0.5, 992.0, 0.1]
+    [0.5, -0.8, 0.1]
   ],
   "detector_sample_summing": 1,
   "detector_line_summing": 1,
-  "t0_ephemeris": -8.0,
+  "t0_ephemeris": -0.8,
   "dt_ephemeris": 0.2,
-  "t0_quaternion": -8.0,
+  "t0_quaternion": -0.8,
```
Comment (jessemapel) [inline tests/data/constVelocityLineScan.json]: All of these are just errors in the original meta-data.

File: tests/data/constVelocityLineScan.json [hunk 5]
```diff
@@ -23,7 +23,7 @@
   "image_samples": 16,
   "detector_center" : {
     "line" : 0.5,
-    "sample" : 7.5
+    "sample" : 8.0
```
Comment (jessemapel) [inline tests/data/constVelocityLineScan.json]: Our CCD is 1 line by 16 samples, so the middle should be 16/2=8, 1/2=0.5

**Comment by jessemapel** [comment]
This changes the ISD spec meaning. detector_center is now expected to be CSM, 0 based pixels. The top left corner of the CCD is (0,0) and pixels are [n,n+1) X [m, m+1).

File: src/UsgsAstroLsSensorModel.cpp (hunk 1 above)
Comment (jlaura) [reply src/UsgsAstroLsSensorModel.cpp]: :+1: no need for f(x_{1})f(x_{2}) < 0 for secant.

File: src/UsgsAstroLsSensorModel.cpp (hunk 2 above)
Comment (jlaura) [reply src/UsgsAstroLsSensorModel.cpp]: I think that we want it in the parameterVals? If we want to do an adjustment and let the focal length vary, we need to be able to make modifications to the adjustments parameter. I think the only way to vary focal is if it is in the parameterVals vector.

File: tests/CMakeLists.txt (hunk 3 above)
Comment (jlaura) [reply tests/CMakeLists.txt]: https://cmake.org/cmake/help/v3.13/command/file.html?#filesystem
> 
> Note
> 
> We do not recommend using GLOB to collect a list of source files from your source tree. If no CMakeLists.txt file changes when a source is added or removed then the generated build system cannot know when to ask CMake to regenerate. The CONFIGURE_DEPENDS flag may not work reliably on all generators, or if a new generator is added in the future that cannot support it, projects using it will be stuck. Even if CONFIGURE_DEPENDS works reliably, there is still a cost to perform the check on every rebuild.

File: tests/LineScanCameraTests.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+   // EXPECT_NO_THROW(
+   //       sensorModel->replaceModelState(modelState)
+   // );
+
+   // When this is different, the output is very hard to parse
+   // TODO implement JSON diff for gtest
+
+   // EXPECT_EQ(sensorModel->getModelState(), modelState);
+}
+
```
Comment (jlaura) [inline tests/LineScanCameraTests.cpp]: what are these?

File: src/UsgsAstroLsSensorModel.cpp [hunk 7]
```diff
@@ -561,57 +528,89 @@ csm::ImageCoord UsgsAstroLsSensorModel::groundToImage(
...
+      auto referenceTimeIt = std::upper_bound(m_intTimeStartTimes.begin(),
+                                              m_intTimeStartTimes.end(),
+                                              nextTime);
+      if (referenceTimeIt != m_intTimeStartTimes.begin()) {
+         --referenceTimeIt;
+      }
+      size_t referenceIndex = std::distance(m_intTimeStartTimes.begin(), referenceTimeIt);
+      double computedLine = (nextTime - m_intTimeStartTimes[referenceIndex]) / m_intTimes[referenceIndex]
+                          + m_intTimeLines[referenceIndex] - 0.5; // subtract 0.5 for ISIS -> CSM pixel conversion
+      double closestLine = floor(computedLine + 0.5);
```
[the remark this answers is not in the document]
Comment (jessemapel) [reply src/UsgsAstroLsSensorModel.cpp]: We should also pull the conversion in a different PR. At the same time we should change the starting sample/line to 0 based.

[the remark this answers is not in the document]
Comment (jessemapel) [reply src/UsgsAstroLsSensorModel.cpp]: Because that will need to be coordinated with pfefferneuse

File: src/UsgsAstroLsSensorModel.cpp [hunk 8]
```diff
@@ -2504,9 +2510,10 @@ csm::ImageCoord UsgsAstroLsSensorModel::computeViewingPixel(
 
    // Convert to image sample line
    double line = detectorLine + m_detectorLineOrigin - m_detectorLineOffset
-               - m_offsetLines + 0.5;
-   double sample = (detectorSample + m_detectorSampleOrigin - m_startingSample)
-                 / m_detectorSampleSumming - m_offsetSamples + 0.5;
+               - m_offsetLines;
+   double sample = (detectorSample + m_detectorSampleOrigin - m_startingSample + 1.0)
```
[the remark this answers is not in the document]
Comment (jessemapel) [reply src/UsgsAstroLsSensorModel.cpp]: Yes see [1694](https://github.com/USGS-Astrogeology/usgscsm/blob/a78ad02ec7b90395a3c1b061e8f517dd139870a0/src/UsgsAstroLsSensorModel.cpp#L1694)

File: src/UsgsAstroLsSensorModel.cpp (hunk 2 above)
Comment (jessemapel) [reply src/UsgsAstroLsSensorModel.cpp]: There's 2 things we should do then.

1. Remove explicit focal length stuff from our metadata stuff.
2. Remove m_focal from all calculations and use adjusted parameter 15.

I will do 2 on this PR.

File: tests/LineScanCameraTests.cpp (hunk 6 above)
Comment (jessemapel) [reply tests/LineScanCameraTests.cpp]: I copied tests over from the framer. The state test should work. There's something wrong with how we are writing out and setting state in the line scanner.

The other tests just need their values re-calculated for the line scan test case.

File: tests/CMakeLists.txt (hunk 3 above)
Comment (jessemapel) [reply tests/CMakeLists.txt]: Is there a better way to manage the list of test sources? This won't be an issue here, but it will be for ISIS.

**Comment by jlaura** [comment]
@jessemapel @kberryUSGS Is this good to merge?  If so, this should auto-rebuild the .dll for @thareUSGS via appveyor.

**Comment by kberryUSGS** [comment]
:+1: This looks good to me.

**Comment by jessemapel** [comment]
This should be ready
